Stage 1 frontend api

rudy5004 opened the pull request:
In Stage 1.1, I set up the frontend of my application using Vite and organized my components into functional, reusable pieces. I implemented the essential routes, designed the layout with responsive techniques like flex and grid, and followed the BEM methodology for CSS class naming. In Stage 1.2, I connected the app to a third-party API, using fetch to make GET and POST requests while handling API errors and dynamically displaying data in the UI. I showed a preloader during data fetching, and implemented a "Show more" button to load additional content. I used local storage to maintain state between visits, and simulated the "Save" button functionality for now.

rudy5004 committed:
stage-1-1-markup- vite setup and stage-react-api branch created

rudy5004 committed:
stage-1-1-markup home page sign in and sign up modal completed

rudy5004 committed:
stage-1-1-markup registration successful modal completed

rudy5004 committed:
stage-1-1-markup saved news page with header and footer

rudy5004 committed:
stage-1-1-markup saved-news card layout

rudy5004 committed:
stage-1-1-markup card rendered successfully

rudy5004 committed:
stage-1-1-markup saved news completed

rudy5004 committed:
stage-1-1-markup data passed to NewsCards

rudy5004 committed:
stage-1-1-markup front end complete except for page breaks

rudy5004 committed:
stage-1-1-markup header page break 720 with functional sign in button

rudy5004 committed:
stage-1-1-markup header page break 720 320 with bar icon functional

rudy5004 committed:
stage-1-1-markup header page break 720 and 320px with black box functionality

rudy5004 committed:
stage-1-1-markup header page break 720 and 320px with black box functionality ans overlay sign in close

rudy5004 committed:
stage-1-1-markup header page break 720 and 320 all but X icon

rudy5004 committed:
stage-1-1-markup / page done and savedNewsHeader done except a few edits

rudy5004 committed:
stage-1-1-markup / page done and savedNewsHeader done except a few edits

rudy5004 committed:
stage-1-1-markup / page done and savedNewsHeader done except X icon

rudy5004 committed:
stage-1-1-markup / page done and savedNewsHeader done

rudy5004 committed:
stage-1-1-markup stage one submission

rudy5004 committed:
stage-1-2-api api setup successfully for / page

rudy5004 committed:
stage-1-2-api api setup successfully for / page with savedNews.svg icon working

rudy5004 committed:
stage-1-2-api submission

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Clay, 

unfortunately I have to reject the project without review:



* Please, **deploy** the app
* In ` README.md` you need to add some info about your project: technologies, how to download, how to start, how to deploy the app. And you need to add the link to your **deployed app**.

rudy5004 committed:
stage-1-2-api submission #2

rudy5004 committed:
stage-1-2-api submission #3

rudy5004 commented:
Thank you for your professional feedback.  I updated the README.md as well as deploying the application via github pages.  The link is provided within the README.md per your request.

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Clay, 

unfortunately I have to reject the project without review:


* There are warnings about `non-unique` **ids** which may cause **bugs** https://snipboard.io/Nm1GLF.jpg 
* The deployed app should be able to search articles. Now it's not allowed. See the errror https://snipboard.io/nwX3vY.jpg 
* There should be a navigation link to the `Saved articles` page where you need to show how the page looks like (you can add hardcoded articles there)
* Please, make the page match the design on 320px

rudy5004 committed:
stage-1-2-api fixed warnings

rudy5004 committed:
stage-1-2-api fixed sign in v2

rudy5004 committed:
stage-1-2-api warnings and public api search fix

rudy5004 committed:
stage-1-2-api saved articles setup

rudy5004 committed:
stage-1-2-api saved articles setup with local Storage

rudy5004 committed:
stage-1-2-api 320px changes top saved-news

rudy5004 committed:
stage-1-2-api saved-news endpoint trashcan icon

rudy5004 committed:
stage-1-2-api Mobile_SavedNews_Logged_in done

rudy5004 commented:
made appropriate changes per your request.  Please let me know if I missed anything.  Again thank you for your time and hard work

gennady-bars commented on src/components/App/App.css:
```diff
@@ -0,0 +1,23 @@
+  min-height: 100vh;
+}
+
+.app main {
```
`.app main {
`

According to BEM, it's not allowed to use **tags** in `css` for styling. You need to add **class names** for the elements

gennady-bars commented on src/components/App/App.css:
```diff
@@ -0,0 +1,23 @@
+/* App.css */
+body {
```
According to BEM, it's not allowed to use **tags** in `css` for styling. You need to add **class names** for the elements 


**Please, correct it everywhere in the code**

gennady-bars commented on src/components/ModalWithForm/ModalWithForm.jsx:
```diff
@@ -0,0 +1,23 @@
+
+function ModalWithForm({ isOpen, onClose, onSubmit, title, children }) {
+  return (
+    <div className={`modal ${isOpen ? "modal_open" : ""}`}>
```
**COULD BE IMPROVED**

If it’s interesting for you here is a way to make a special wrapper for any `modal` with cross button, `Escape` and `overlay` listeners inside. You should create component `Modal`, store it in `Modal.js`. 


```jsx

import { useEffect } from "react";

export const Modal = ({ name, onClose, children }) => {
  // here is `useEffect` for the `Escape` listener
  useEffect(() => {
    // we should define the handler inside `useEffect`, so that it wouldn’t lose the reference to be able to remove it
    const handleEscape = (e) => {
      if (e.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    // don’t forget to remove the listener in the `clean-up` function
    return () => document.removeEventListener("keydown", handleEscape);
  }, [onClose]);

  // here is the overlay handler
  const handleOverlay = (e) => {
    if (e.target === e.currentTarget) {
      onClose();
    }
  };

  // then we add the main wrapper with class `modal`
  return (
    <div className={`modal modal_type_${name}`} onClick={handleOverlay}>
      {/* the container for the contents */}
      <div className="modal__container">
        {/* here will be anything you add as `children`*/}
        {children}
        {/* add the close button */}
        <button className="modal__close" type="button" onClick={onClose} />
      </div>
    </div>
  );
};



```


And now you can use `Modal` for any popup in the project:  `ItemModal`  and `ModalWithForm` 

```jsx


function ModalWithForm({ name, onClose, ...props}) {
  return (
    <Modal name={name} onClose={onClose}>
        <h2 className='modal__title'>{props.title}</h2>


```

You can make also `Input` for inputs and `Form` for any form in the project.

gennady-bars commented on src/components/SignUpModal/SignUpModal.jsx:
```diff
@@ -0,0 +1,143 @@
+  }, [email, username, password]);
+
+  return (
+    <>
```
`<>` empty tags are not needed if you have a common tag that wraps the whole markup of the component

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Clay!

Great job on submitting your project! 

**Parts I really liked in your project:**


* Well-structured markup
* No validation errors in `css`
* All `img` tags have descriptive `alt` attributes
* Great semantic tags
* Good job that submit handlers are added only to the `form` tag
* Good job that you `catch` possible errors at the end of server requests




There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:
* According to BEM, it's not allowed to use **tags** in `css` for styling. You need to add **class names** for the elements 
*  Please, make the `Saved articles` navigation link match the design https://snipboard.io/v84Ep5.jpg 
*  Please, make the `Show more` button match the design  https://snipboard.io/rYo9uf.jpg
* Please, make the items match the design at the bottom if there is a lot of text https://snipboard.io/8AmGWE.jpg 
* There is a **horizontal** scrolling on the Saved articles page https://snipboard.io/UCDGrs.jpg 
* Please, show the `exit` button with the name of the user on the `home` page after logging in 
* Please, add `:hover` effects for the **navigation** links
* There should be a link for `login` in the sign up modal https://snipboard.io/qvjkA9.jpg 
* `<>` empty tags are not needed if you have a common tag that wraps the whole markup of the component

## Number of points: 89


# Summary:

### Performance Criteria

#### Markup and JSX

- [] <!--7.4--> The project is adapted for different screen resolutions and looks correct on all intermediate resolutions. There is no horizontal scrolling on resolutions of `320px` and up. The scrollbar can't be hidden using the overflow: hidden declaration.
- [x] <!--7.26--> Navigation between pages and links to external resources are working correctly. There are no broken links or links leading to an empty anchor on the page, and external links open in a new tab. All project pages can be accessed, and there are no hidden blocks.
- [x] <!--7.26--> Elements specified in the UI toolkit are used correctly (e.g., buttons, forms, text elements, etc).
- [] <!--3.63--> Classes are named according to BEM specifications.
- [x] <!--3.63--> Semantic HTML is used, meaning that semantic tags are used. All elements are used correctly (e.g., a paragraph must be a paragraph, a list must be a list). The DOM tree structure doesn't consist only of `<div />` containers.
- [x] <!--3.63--> The correct approach for positioning elements has been chosen and is described using the correct syntax. For example, when positioning an element absolutely, its parent block is relatively positioned.
- [x] <!--3.63--> `flex` or `grid` layouts are used to arrange elements whenever applicable.
- [x] <!--3.63--> Infrastructural project files are created using CRA.
- The markup has been converted into JSX:
  - [x] <!--3.63--> The markup is included within parentheses `( )`.
  - [x] <!--3.63--> The markup has been moved into the corresponding components.
- [x] <!--3.63--> Your project contains:
  - [] An `images`
  - [x] A `components` folder with JS and CSS files for the components.
  - [x] A `fonts` folder.
- [x] <!--3.63--> There are no linting errors or warnings.

#### React

- React components:
  - [x] <!--3.63--> Hooks are not used inside conditional statements or loops.
  - [x] <!--3.63--> Hooks are called in a component's main function.
  - [x] <!--3.63--> For class components, effects are described inside the component lifecycle methods.
- [x] <!--7.26--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--3.63--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--3.63--> No third-party JavaScript libraries are used, unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

#### Markup and JSX

- [x] <!--1.25--> The design reuses components wherever possible.
- [x] <!--1.25--> Fonts are connected using `@font-face`.
- [x] <!--1.25--> The icons are exported in SVG format.
- [x] <!--1.25--> Form elements should be highlighted when focused.
- [x] <!--1.25--> The form must have placeholders,  and there are required properties for the fields.
- [x] <!--1.25--> The use of `reset.css` is prohibited.

#### React

- [x] <!--1.25--> The initial state of state variables contain the correct data type.
- [x] <!--1.25--> API requests are described inside the `App` component.
- [x] <!--1.25--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [x] <!--1.25--> API error handling:
  - [x] The user receives a message in the case of an error.
- [x] <!--1.25--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.
- [x] <!--1.25--> If you are using popups, then correct the opening and closing functionality is implemented. Popups can be closed using the cross button, overlay, or ESC key.

### Recommendations

#### Markup and JSX

- [x] <!--0.83--> System fonts are connected as alternatives to each of your fonts.
- [x] <!--0.83--> Images have an `alt` attribute containing appropriate values.
- [x] <!--0.83--> Raster and vector images have been optimized.

#### React

- [x] <!--0.83--> For internal links in the application, components from the react-router library are used.
- [x] <!--0.83--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--0.85--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points: 89

## Additional comments:

-

rudy5004 committed:
stage-1-2-api removed <> empty tags

rudy5004 committed:
stage-1-2-api show more button fixed

rudy5004 committed:
stage-1-2-api horizontal scrolling fixed

rudy5004 committed:
stage-1-2-api username displayed button on  savedarticles page

rudy5004 committed:
stage-1-2-api savedarticles css alignment fixed

rudy5004 committed:
stage-1-2-api header sign out button with username fixed

rudy5004 committed:
stage-1-2-api login sign up button fixed

rudy5004 committed:
stage-1-2-api fixed hyperlinks

rudy5004 committed:
stage-1-2-api fixed BEM for App.jsx

rudy5004 committed:
stage-1-2-api fixed BEM for Footer.jsx

rudy5004 committed:
stage-1-2-api fixed all other BEM .jsx components

rudy5004 commented:
Hello thank you for the feedback.  I have made the following changes:

- Made appropriate BEM changes
- Made the Saved Articles Navigation link match the Figma
- Made the show more button match the Figma
- Corrected the NewsCard Text at the bottom
- Fixed Horizontal Scrolling
- Created and styled the exit button with the name of the user on the home page after logging in
- added the login link to the sign up modal
- removed <> empty tags per components

gennady-bars reviewed (commented):
Everything is almost fine, but:


* Please, make the backgound match the design when the articles appear on the `home` page https://snipboard.io/pj930M.jpg 
*  One remark hasn't been corrected https://github.com/rudy5004/headline_news_frontend/pull/1/files#r1847457866  According to BEM, it's not allowed to use **tags** in `css` for styling. You need to add **class names** for the elements 
* Please, add `:hover` effects for the **navigation** links https://snipboard.io/7WnAQO.jpg

rudy5004 committed:
stage-1-2-api edit background color for home page

rudy5004 committed:
stage-1-2-api edit hover effects for all navigation links

rudy5004 committed:
stage-1-2-api fixed BEM noncompliance with Preloader.jsx

rudy5004 commented:
Thanks for your help.  I have corrected the following:

-background on home page
-hover effects for navigation links
-BEM compliance fix

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
